# ETL Notebook 
This notebook documents my step-by-step thinking and justification behind my decisions for building the ETL pipeline.

## EXTRACT

In [108]:
from pathlib import Path
import pandas as pd
import ast

Each athlete lives in a 'p[XX]/pmsys/' folder. I glob 'p[0-9][0-9]' so the loop picks up only athlete folders and never the 'participant-overview.xlsx', and 'sorted()' makes athlete order deterministic. This way p01 always processes first, so generated IDs are reproducible across runs.

Two absence cases are handled on differently on purpose. p08 has no 'injury.csv' file at all, so the injury read is guardxed and p08 simply contributes zero injury rows (a missing file is valid, not an error). 'srpe.csv' and 'wellness.csv' are present for every athlete, so those areads are intentionally unguarded. Missing on of these files should indicate that the extract has failed.

In [109]:
PMDATA_DIR = Path("data/pmdata")

injury_frames, srpe_frames, wellness_frames = [], [], []

# p[0-9][0-9] matches p01..p16 only — NOT participant-overview.xlsx
# sorted() keeps them in order
for athlete_dir in sorted(PMDATA_DIR.glob("p[0-9][0-9]")):
    source_id = athlete_dir.name # e.g. "p01"
    pmsys = athlete_dir / "pmsys"

    # srpe + wellness are present for every athlete
    srpe = pd.read_csv(pmsys / "srpe.csv", parse_dates=["end_date_time"])
    srpe.insert(0, "source_id", source_id)
    srpe_frames.append(srpe)

    wellness = pd.read_csv(pmsys / "wellness.csv", parse_dates=["effective_time_frame"])
    wellness.insert(0, "source_id", source_id)
    wellness_frames.append(wellness)

    # guard: p08 has no injury.csv
    injury_path = pmsys / "injury.csv"
    if injury_path.exists():
        injury = pd.read_csv(injury_path, parse_dates=["effective_time_frame"])
        injury.insert(0, "source_id", source_id)
        injury_frames.append(injury)
    else:
        print(f"skipping injury for {source_id}: {injury_path} not found")

injury = pd.concat(injury_frames, ignore_index=True)
srpe = pd.concat(srpe_frames, ignore_index=True)
wellness = pd.concat(wellness_frames, ignore_index=True)

for name, df in {"injury": injury, "srpe": srpe, "wellness": wellness}.items():
    print(f"{name}: {df.shape[0]} rows x {df.shape[1]} cols, "
          f"{df['source_id'].nunique()} athletes")

skipping injury for p08: data\pmdata\p08\pmsys\injury.csv not found
injury: 225 rows x 3 cols, 15 athletes
srpe: 783 rows x 5 cols, 16 athletes
wellness: 1747 rows x 10 cols, 16 athletes


Sanity check on participant_overview df

In [110]:
participant_overview = pd.read_excel(PMDATA_DIR / "participant-overview.xlsx", header=1)
participant_overview

,Participant ID,Age,Height,Gender,A or B person,Max heart rate,Date,Minutes,Seconds,Stride walk,Stride run
0,p01,48,195,male,A,182,2019-11-26 00:00:00,29,33,80.90,102.9
1,p02,60,180,male,A,169,2019-12-15 00:00:00,23,51,74.70,92.4
2,p03,25,184,male,A,157,2019-12-30 00:00:00,33,22,NaN,NaN
3,p04,26,163,female,A,195,2019-11-19 00:00:00,22,13,67.30,110.2
4,p05,35,176,male,A,184,2019-12-23 00:00:00,32,40,73.00,94.3
5,p06,42,179,male,B,181,2019-12-01 00:00:00,23,19,73.04,97.6
6,p07,26,177,male,B,NaN,2019-11-19 00:00:00,19,40,73.50,119.5
7,p08,27,186,male,B,200,2019-11-28 00:00:00,18,47,77.20,103.6
8,p09,26,180,male,B,183,2020-01-07 00:00:00,35,6,74.70,109.9
9,p10,38,179,female,B,197,2019-12-08 00:00:00,28,10,73.09,102.3


Decided to only retain demographic information about each athlete:

In [111]:
participant_overview = participant_overview.drop(columns=['Date', 'Minutes', 'Seconds', 'Stride walk', 'Stride run'])
participant_overview

,Participant ID,Age,Height,Gender,A or B person,Max heart rate
0,p01,48,195,male,A,182
1,p02,60,180,male,A,169
2,p03,25,184,male,A,157
3,p04,26,163,female,A,195
4,p05,35,176,male,A,184
5,p06,42,179,male,B,181
6,p07,26,177,male,B,NaN
7,p08,27,186,male,B,200
8,p09,26,180,male,B,183
9,p10,38,179,female,B,197


I noticed p08 had no injury.csv file and wanted to inspect how many injuries were being reported and how many of them were empty per athlete:

In [112]:
injury["is_empty"] = injury["injuries"].isin([{}, "{}"])   # dict vs string, see below
print(injury.groupby("source_id")["is_empty"].agg(["sum", "count"]))

           sum  count
source_id            
p01         23     24
p02          4      5
p03         11     11
p04         14     17
p05          0     10
p06         15     15
p07         13     15
p09         20     20
p10         12     12
p11          1      4
p12          1     49
p13         16     16
p14          7     10
p15          0      6
p16         11     11


This tells us that there aren't clean periodic check-ins. Otherwise we would have "{}" - empty injury check-ins on days where the athlete is healthy and injury free.  

Next, I want to see if there are persisting injuries being logged repeatedly. Let's take p12 for example and see what injuries are being reported:

In [113]:
p12 = injury[injury["source_id"] == "p12"]
p12[p12["injuries"] != "{}"][["effective_time_frame", "injuries"]].head(20)

,effective_time_frame,injuries
133,2019-11-15 20:01:20.647000+00:00,{'right_knee': 'minor'}
135,2019-11-18 04:50:00.666000+00:00,{'right_knee': 'minor'}
136,2019-11-18 06:01:40.574000+00:00,{'right_knee': 'minor'}
137,2019-11-18 19:26:24.936000+00:00,{'left_shoulder': 'minor'}
138,2019-11-19 05:49:39.091000+00:00,"{'left_leg': 'minor', 'left_shoulder': 'minor'..."
139,2019-11-20 05:58:09.329000+00:00,"{'left_leg': 'minor', 'left_shoulder': 'minor'..."
140,2019-11-20 20:03:10.579000+00:00,"{'left_leg': 'minor', 'right_knee': 'minor'}"
141,2019-11-21 20:59:32.959000+00:00,"{'left_leg': 'minor', 'right_knee': 'minor'}"
142,2019-11-22 22:35:26.660000+00:00,"{'left_leg': 'minor', 'right_knee': 'minor'}"
143,2019-11-24 10:10:14.803000+00:00,{'right_knee': 'minor'}


Note that there are multiple injuries being reported on the same day. These will need to be exploded so that a single row turns into N-number (number of injured areas per day) of rows so there is one per body part, each with that part's severity.

## TRANSFORM

### Injury: grain and the report-vs-episode decision

A single row in the 'injury.csv' file is a one weekly check-in, not one injury. '{}' means that "checked in, no injury" which is useful information indicating the athlete was healthy that week. Hence, these rows don't belong in an injury fact table. Non-empty rows hold a '{body_part: severity}' dict, and one row can name several body parts, so I explode to one row per (athlete, timestamp, body_part, severity).

Key decision: I chose to keep the full timestamp rather than collapsing to a date. Several athletes re-report the *same* injury across consecutive check-ins (p12's right knee spans weeks). Collapsing to a date would manufacture false duplicates and, worse, counting rows would wildy overstate injury counts. So the fact table stays at report grain which is lossless and faithful to the source. The "collapse reapeted reports into injury episodes" is left as a downstream derivation, not baked into the load.

'injuries' arrives from 'read_csv' as a string and not a dict. I parse with 'ast.literal_eval' and confirm every row parses to a dict before exploding.

In [114]:
injury = injury.drop(columns="is_empty") # Removing the checker column that was previously created
injury["injuries"] = injury["injuries"].apply(ast.literal_eval) # Convert injuries column to dicts


In [115]:
injured = injury[injury["injuries"] != {}]
injured

,source_id,effective_time_frame,injuries
11,p01,2020-01-07 22:33:38.989000+00:00,{'right_hand': 'minor'}
24,p02,2019-11-13 13:53:01.675000+00:00,"{'left_elbow': 'minor', 'right_shoulder': 'min..."
43,p04,2019-11-26 06:24:57.956000+00:00,{'groin_hip': 'minor'}
50,p04,2020-01-08 08:03:03.131000+00:00,"{'left_leg': 'minor', 'right_leg': 'minor'}"
53,p04,2020-01-26 21:34:24.956000+00:00,"{'left_leg': 'minor', 'right_leg': 'minor'}"
...,...,...,...
209,p15,2020-02-10 08:05:01.763000+00:00,{'stomach_back': 'major'}
210,p15,2020-02-11 08:10:36.015000+00:00,{'stomach_back': 'minor'}
211,p15,2020-02-13 05:46:38.880000+00:00,{'stomach_back': 'major'}
212,p15,2020-02-15 07:22:56.107000+00:00,{'stomach_back': 'major'}


In [116]:
records = []
for row in injured.itertuples():
    ath_id = row.source_id
    effective_time_frame = row.effective_time_frame
    inj = row.injuries
    for body_part, severity in inj.items():
        records.append({
            "source_id": ath_id,
            "effective_time_frame": effective_time_frame,
            "body_part": body_part,
            "severity": severity,
        })
injuries_fact = pd.DataFrame(records)

In [117]:
injuries_fact.shape

(111, 4)

In [118]:
injuries_fact["body_part"].value_counts()

body_part
right_knee        35
left_foot         15
left_leg          11
right_foot         9
left_knee          5
left_hand          5
left_shoulder      5
stomach_back       5
chest              4
right_hand         4
head_neck          3
right_leg          3
groin_hip          2
right_shoulder     2
left_elbow         1
right_thigh        1
right_elbow        1
Name: count, dtype: int64

In [119]:
injuries_fact

,source_id,effective_time_frame,body_part,severity
0,p01,2020-01-07 22:33:38.989000+00:00,right_hand,minor
1,p02,2019-11-13 13:53:01.675000+00:00,left_elbow,minor
2,p02,2019-11-13 13:53:01.675000+00:00,right_shoulder,minor
3,p04,2019-11-26 06:24:57.956000+00:00,groin_hip,minor
4,p04,2020-01-08 08:03:03.131000+00:00,left_leg,minor
...,...,...,...,...
106,p15,2020-02-10 08:05:01.763000+00:00,stomach_back,major
107,p15,2020-02-11 08:10:36.015000+00:00,stomach_back,minor
108,p15,2020-02-13 05:46:38.880000+00:00,stomach_back,major
109,p15,2020-02-15 07:22:56.107000+00:00,stomach_back,major


Once exploded, (source_id, timestmap, body_part) is unique - verified with a duplicate check returning 0. I still add a surrogate 'injury_id' as the primary key: a single-column integer key gives clean joins, and the natural composite key becoems a UNIQUE constraint that documents the grain guarantee.

Adding a surrogate key for injury (injury_id):

In [120]:
injuries_fact = injuries_fact.reset_index(drop=True)
injuries_fact.insert(0, "injury_id", range(1, len(injuries_fact) + 1))
injuries_fact

,injury_id,source_id,effective_time_frame,body_part,severity
0,1,p01,2020-01-07 22:33:38.989000+00:00,right_hand,minor
1,2,p02,2019-11-13 13:53:01.675000+00:00,left_elbow,minor
2,3,p02,2019-11-13 13:53:01.675000+00:00,right_shoulder,minor
3,4,p04,2019-11-26 06:24:57.956000+00:00,groin_hip,minor
4,5,p04,2020-01-08 08:03:03.131000+00:00,left_leg,minor
...,...,...,...,...,...
106,107,p15,2020-02-10 08:05:01.763000+00:00,stomach_back,major
107,108,p15,2020-02-11 08:10:36.015000+00:00,stomach_back,minor
108,109,p15,2020-02-13 05:46:38.880000+00:00,stomach_back,major
109,110,p15,2020-02-15 07:22:56.107000+00:00,stomach_back,major


### Wellness: dropping empty submissions

The grain for the wellness csv is different to the injury one. Some athletes submit wellness more than once a day, and inspection shows these are genuinely different reports (morning vs evening values differ), not duplications. So (athlete, day) is not an appropriate grain and would cause the same mistake as collapsing injury to a date. I store welness at report grain, keyed on (athlete_id, reported_at), and defer the one-row-per-day collapse to a SQL view (first report of day). Cleaning stays reversible and the base table stays faithful. 

I want to examine if there are multiple wellness reports on the same day:

In [121]:
wellness["d"] = wellness["effective_time_frame"].dt.date
multi = wellness.groupby(["source_id","d"]).filter(lambda g: len(g) > 1)
multi.sort_values(["source_id","d","effective_time_frame"])

,source_id,effective_time_frame,fatigue,mood,readiness,sleep_duration_h,sleep_quality,soreness,soreness_area,stress,d
257,p03,2019-12-01 07:38:02.291000+00:00,3,3,5,6,3,3,[],3,2019-12-01
258,p03,2019-12-01 21:12:20.093000+00:00,3,3,5,5,3,3,[],3,2019-12-01
261,p03,2019-12-11 06:41:11.728000+00:00,3,3,5,7,3,3,[],3,2019-12-11
262,p03,2019-12-11 17:01:26.334000+00:00,3,3,5,7,3,3,[],3,2019-12-11
278,p03,2020-01-12 00:32:08.115000+00:00,3,3,5,7,3,3,[],3,2020-01-12
279,p03,2020-01-12 15:28:39.900000+00:00,3,3,5,7,3,3,[],3,2020-01-12
297,p03,2020-02-02 01:00:28.003000+00:00,3,3,5,8,3,3,[],3,2020-02-02
298,p03,2020-02-02 18:15:39.697000+00:00,3,3,5,8,3,3,[],3,2020-02-02
439,p04,2020-03-07 08:17:42.897000+00:00,3,3,5,8,4,3,[],3,2020-03-07
440,p04,2020-03-07 08:18:13.008000+00:00,3,3,5,8,3,3,[],3,2020-03-07


This is an interesting find since the paper states that fatigue, mood, sleep_quality, soreness and stress are rated on a 1-5 scale while readiness is rated 0-10. A row where all five are 0 is therefore an empty/aborted submission, not a real report. I drop only those fully-empty rows (`.all(axis=1)`, not `.any()`): rows missing a single field are genuine reports with one blank value and are kept. Verified this removes exactly the expected rows.

In [122]:
five_pt_cols = ['fatigue', 'mood', 'sleep_quality', 'soreness', 'stress']
wellness[wellness[five_pt_cols].eq(0).all(axis=1)]

,source_id,effective_time_frame,fatigue,mood,readiness,sleep_duration_h,sleep_quality,soreness,soreness_area,stress,d
502,p05,2019-12-12 10:50:32.079000+00:00,0,0,0,0,0,0,[],0,2019-12-12
510,p05,2019-12-20 06:58:36.783000+00:00,0,0,0,0,0,0,[],0,2019-12-20


In [123]:
empty_mask = wellness[five_pt_cols].eq(0).all(axis=1)
wellness = wellness[~empty_mask]
wellness = wellness.reset_index(drop=True)

Now I want to check if any of these columns have a 0 to see how to handle this case:

In [124]:
five_pt_cols = ['fatigue', 'mood', 'sleep_quality', 'soreness', 'stress']
wellness[wellness[five_pt_cols].eq(0).any(axis=1)]

,source_id,effective_time_frame,fatigue,mood,readiness,sleep_duration_h,sleep_quality,soreness,soreness_area,stress,d
373,p04,2019-12-27 09:00:21.074000+00:00,2,3,3,6,2,1,"[48979004, 32696007, 12921003, 22943007, 36820...",0,2019-12-27
1201,p11,2019-12-06 08:39:06.697000+00:00,2,0,5,6,3,3,[],2,2019-12-06
1406,p13,2019-12-25 17:05:13.971000+00:00,2,3,3,6,2,2,[12921003],0,2019-12-25


Some fields are full but others are at 0. This tells me that the rows are still real reports and the athlete has just not logged 'stress' or 'mood'. So I chose to drop rows where all five are 0 but treated a lone zero as a missing value and a reason to not drop the entire row.

---

### Training sessions: dropping empty submissions and splitting the activity list

'activity_names' blends two different concepts in one list e.g. '['individual', 'strength']'. It consists of participation *mode* with an activity *type*. I split them: exactly one mode per session (verified across all sessions) becomes a scalar 'participation_mode' column on 'training_session', while the 1-to-many activity types explode into a separate 'session_activity' bridge table. 'session_load' is materialized here as RPE x duration - a derived column I store deliberately because it's the workload primitive ACWR sums over, and I'd rather compute it once than recompute it in every window query.
 
 Next, I want to see if the participation mode ("individual" or "team") is being reported once per activity in the srpe.csv file: 

In [125]:
srpe["parsed"] = srpe["activity_names"].apply(ast.literal_eval)
srpe["parsed"].apply(lambda l: len([x for x in l if x in {"individual","team"}])).value_counts()

parsed
1    772
0     11
Name: count, dtype: int64

772 rows have the participation mode in the activity name and 11 don't. I also want to see the data type distribution and check for the entries that don't have participation mode:

In [126]:
srpe["parsed"].apply(type).value_counts()

parsed
<class 'list'>    772
<class 'dict'>     11
Name: count, dtype: int64

In [127]:
srpe[srpe["parsed"].apply(type) == dict][["end_date_time", "activity_names", "parsed"]]

,end_date_time,activity_names,parsed
772,2019-11-18 13:34:02.327000+00:00,{},{}
773,2019-11-25 14:25:04.577000+00:00,{},{}
774,2019-12-09 12:32:35.414000+00:00,{},{}
775,2019-12-23 16:12:08.718000+00:00,{},{}
776,2019-12-30 10:29:31.987000+00:00,{},{}
777,2020-01-13 15:37:07.243000+00:00,{},{}
778,2020-01-20 15:42:00.634000+00:00,{},{}
779,2020-01-28 20:53:03.947000+00:00,{},{}
780,2020-02-04 21:43:29.297000+00:00,{},{}
781,2020-02-10 15:29:39.385000+00:00,{},{}


In [128]:
srpe[srpe["parsed"].apply(type) == dict][["perceived_exertion", "duration_min", "parsed"]]

,perceived_exertion,duration_min,parsed
772,NaN,NaN,{}
773,NaN,NaN,{}
774,NaN,NaN,{}
775,NaN,NaN,{}
776,NaN,NaN,{}
777,NaN,NaN,{}
778,NaN,NaN,{}
779,NaN,NaN,{}
780,NaN,NaN,{}
781,NaN,NaN,{}


11 srpe rows have an empty `{}` activity *and* NaN RPE and NaN duration - no activity, no load, nothing computable. Unlike injury `{}` (which carried information), these carry none, so I drop them. I filter on "keep rows that parsed to a list" rather than "drop the dict rows. Keeping the known-good shape is safer than blacklisting the one bad shape I happened to find.

In [129]:
before = len(srpe)
srpe = srpe[srpe["parsed"].apply(type) == list]   # keep only the list rows
srpe = srpe.reset_index(drop=True)
print(before - len(srpe))   # expect 11

11


In [ ]:
training_records = []
for row in srpe.itertuples():
    parsed = row.parsed
    mode = [x for x in parsed if x in {"individual", "team"}][0]

    training_records.append({
        "source_id": row.source_id, 
        "end_time": row.end_date_time,
        "rpe": row.perceived_exertion,
        "duration_min": row.duration_min,
        "session_load": row.perceived_exertion * row.duration_min,
        "participation_mode": mode
    })

training_session = pd.DataFrame(training_records)

training_session = training_session.reset_index(drop=True)
training_session.insert(0, "session_id", range(1, len(training_session) + 1))

training_session.head(20)

,session_id,source_id,end_time,rpe,duration_min,session_load,participation_mode
0,1,p01,2019-11-05 22:51:54.710000+00:00,7.0,30.0,210.0,individual
1,2,p01,2019-11-11 21:15:15.092000+00:00,6.0,30.0,180.0,individual
2,3,p01,2019-11-14 21:00:53+00:00,7.0,60.0,420.0,team
3,4,p01,2019-11-18 21:34:44.995000+00:00,6.0,50.0,300.0,individual
4,5,p01,2019-11-21 21:06:25.146000+00:00,6.0,70.0,420.0,team
5,6,p01,2019-11-23 09:34:50.739000+00:00,8.0,30.0,240.0,individual
6,7,p01,2019-11-26 08:03:56.794000+00:00,7.0,40.0,280.0,individual
7,8,p01,2019-11-28 21:07:17.918000+00:00,6.0,60.0,360.0,team
8,9,p01,2019-12-03 08:13:07.418000+00:00,6.0,40.0,240.0,individual
9,10,p01,2019-12-05 22:00:26.647000+00:00,6.0,60.0,360.0,team


'session_activity' rows carry the 'session_id' assigned when I built 'training_session'. Because both frames preserve srpe's order, session_id for row i is i+1. I assert 'len(srpe) == len(training_session)' before relying on that positional alignment. A more robust solution would be to attach session_id to srpe directly so the key travels with the data regardless of order.

In [131]:
assert len(srpe) == len(training_session) # Guard to ensure frames are aligned

activity_records = []
for i, row in enumerate(srpe.itertuples()):
    session_id = i + 1
    parsed = row.parsed
    activities = [x for x in parsed if x not in {"individual", "team"}]
    for activity in activities:
        activity_records.append({
            "session_id": session_id,
            "activity_type": activity,
        })
session_activity = pd.DataFrame(activity_records)
session_activity

,session_id,activity_type
0,1,running
1,2,running
2,3,soccer
3,4,endurance
4,5,soccer
...,...,...
836,768,endurance
837,769,endurance
838,770,endurance
839,771,endurance


In [132]:
session_activity["activity_type"].value_counts()

activity_type
running      402
endurance    257
strength     170
soccer        12
Name: count, dtype: int64

Now I want to create the daily_soreness and body_part lookup:

---

# LOAD

### Athlete dimension and the ID bridge

The athlete dimension is the hub every fact references, so it's built first. It comes from 'participant-overview.xlsx', which has all 16 athletes including p08 who has demographics, wellness and srpe but no injuries. The dimension is the complete roster and the facts are sparse against it.

max_heart_rate arrives with missing values as both blanks and NaNs; I coerce the column withi 'pd.to_numeric(errors="coerce")' so all missing forms become NaN (NULL in SQL). I doin't fill them since NULL is an honest representation of "not measured". I keep both 'athlete_id' (integer surrogate PK) and 'source_id' ("p01" is a UNIQUE natural key) to create a brdige every fact maps through.

In [133]:
participant_overview

,Participant ID,Age,Height,Gender,A or B person,Max heart rate
0,p01,48,195,male,A,182
1,p02,60,180,male,A,169
2,p03,25,184,male,A,157
3,p04,26,163,female,A,195
4,p05,35,176,male,A,184
5,p06,42,179,male,B,181
6,p07,26,177,male,B,NaN
7,p08,27,186,male,B,200
8,p09,26,180,male,B,183
9,p10,38,179,female,B,197


In [134]:
participant_overview["Max heart rate"] = pd.to_numeric(participant_overview["Max heart rate"], errors = "coerce")
participant_overview["Gender"] = participant_overview["Gender"].str.strip()
participant_overview

,Participant ID,Age,Height,Gender,A or B person,Max heart rate
0,p01,48,195,male,A,182.0
1,p02,60,180,male,A,169.0
2,p03,25,184,male,A,157.0
3,p04,26,163,female,A,195.0
4,p05,35,176,male,A,184.0
5,p06,42,179,male,B,181.0
6,p07,26,177,male,B,NaN
7,p08,27,186,male,B,200.0
8,p09,26,180,male,B,183.0
9,p10,38,179,female,B,197.0


In [135]:
athlete = participant_overview.rename(columns={
    'Participant ID': 'source_id',
    'Age': 'age',
    'Height': 'height_cm',
    'Gender': 'gender',
    'A or B person': 'personality_type',
    'Max heart rate': 'max_heart_rate'
    })
athlete = athlete.reset_index(drop=True)
athlete.insert(0, "athlete_id", range(1, len(athlete) + 1))
athlete


,athlete_id,source_id,age,height_cm,gender,personality_type,max_heart_rate
0,1,p01,48,195,male,A,182.0
1,2,p02,60,180,male,A,169.0
2,3,p03,25,184,male,A,157.0
3,4,p04,26,163,female,A,195.0
4,5,p05,35,176,male,A,184.0
5,6,p06,42,179,male,B,181.0
6,7,p07,26,177,male,B,NaN
7,8,p08,27,186,male,B,200.0
8,9,p09,26,180,male,B,183.0
9,10,p10,38,179,female,B,197.0


Checking data type for proper handling:

In [136]:
athlete["personality_type"].unique()

array(['A', 'B'], dtype=object)

In [137]:
id_map = athlete.set_index("source_id")["athlete_id"]
injuries_fact["athlete_id"] = injuries_fact["source_id"].map(id_map)
wellness["athlete_id"] = wellness["source_id"].map(id_map)
training_session["athlete_id"] = training_session["source_id"].map(id_map)

In [138]:
injuries_fact["athlete_id"].isna().sum() # expect 0
wellness["athlete_id"].isna().sum() # expect 0
training_session["athlete_id"].isna().sum() # expect 0

0

In [139]:
injuries_fact

,injury_id,source_id,effective_time_frame,body_part,severity,athlete_id
0,1,p01,2020-01-07 22:33:38.989000+00:00,right_hand,minor,1
1,2,p02,2019-11-13 13:53:01.675000+00:00,left_elbow,minor,2
2,3,p02,2019-11-13 13:53:01.675000+00:00,right_shoulder,minor,2
3,4,p04,2019-11-26 06:24:57.956000+00:00,groin_hip,minor,4
4,5,p04,2020-01-08 08:03:03.131000+00:00,left_leg,minor,4
...,...,...,...,...,...,...
106,107,p15,2020-02-10 08:05:01.763000+00:00,stomach_back,major,15
107,108,p15,2020-02-11 08:10:36.015000+00:00,stomach_back,minor,15
108,109,p15,2020-02-13 05:46:38.880000+00:00,stomach_back,major,15
109,110,p15,2020-02-15 07:22:56.107000+00:00,stomach_back,major,15


In [140]:
wellness = wellness.drop(columns="d")
wellness


,source_id,effective_time_frame,fatigue,mood,readiness,sleep_duration_h,sleep_quality,soreness,soreness_area,stress,athlete_id
0,p01,2019-11-01 08:31:40.751000+00:00,2,3,5,6,3,2,[12921003],3,1
1,p01,2019-11-02 10:00:01.229000+00:00,2,3,6,6,3,2,[12921003],3,1
2,p01,2019-11-03 14:28:03.263000+00:00,3,3,8,6,3,3,[],3,1
3,p01,2019-11-04 07:05:28.429000+00:00,3,3,8,6,3,3,[],3,1
4,p01,2019-11-05 06:13:35.998000+00:00,3,3,8,5,3,3,[],3,1
...,...,...,...,...,...,...,...,...,...,...,...
1740,p16,2020-03-08 11:44:44.398000+00:00,3,3,4,6,2,3,[],3,16
1741,p16,2020-03-11 11:33:15.168000+00:00,2,3,4,6,2,3,[],2,16
1742,p16,2020-03-15 08:33:15.985000+00:00,4,4,4,10,4,4,[],4,16
1743,p16,2020-03-17 09:54:27.989000+00:00,2,3,4,5,3,3,[],3,16


In [141]:
training_session

,session_id,source_id,end_time,rpe,duration_min,session_load,participation_mode,athlete_id
0,1,p01,2019-11-05 22:51:54.710000+00:00,7.0,30.0,210.0,individual,1
1,2,p01,2019-11-11 21:15:15.092000+00:00,6.0,30.0,180.0,individual,1
2,3,p01,2019-11-14 21:00:53+00:00,7.0,60.0,420.0,team,1
3,4,p01,2019-11-18 21:34:44.995000+00:00,6.0,50.0,300.0,individual,1
4,5,p01,2019-11-21 21:06:25.146000+00:00,6.0,70.0,420.0,team,1
...,...,...,...,...,...,...,...,...
767,768,p15,2020-03-24 12:30:25+00:00,6.0,100.0,600.0,individual,15
768,769,p15,2020-03-25 15:30:39+00:00,4.0,90.0,360.0,individual,15
769,770,p15,2020-03-26 13:00:39+00:00,6.0,120.0,720.0,individual,15
770,771,p15,2020-03-27 15:59:02+00:00,3.0,100.0,300.0,individual,15


Every fact carried 'source_id' through transformation. Now that the athlete dimesion defines source_id -> athlete_id, I map each fact's 'soruce_id' to its integer 'athlete_id' to satisfy the foreign keys, and check that no mapped key is null (every fact must reference an athlete that exists). 'source_id' is dropped from teh facts before load and it lives only on the athlete table.

In [142]:
id_map

source_id
p01     1
p02     2
p03     3
p04     4
p05     5
p06     6
p07     7
p08     8
p09     9
p10    10
p11    11
p12    12
p13    13
p14    14
p15    15
p16    16
Name: athlete_id, dtype: int64